### PR TITLE
Unlock Protocol

### PR DESCRIPTION
**Pull request by valentinludu**
Unlock Protocol implementation

Main changes and features:

- Added Unlock to the Token Gate modal
- TokenGate model schema will have a type and conditions can be any json. + Added validation of the payload for POST requests. In the future we will probably have many types of token gates
- Moved mutations from charmClient to hooks/tokenGates for better state management from the swr library
- Use unlock-protocol library to manage lock details
- Payload validation with yup - for conditions

Tested:
✅ Regression testing for lit protocol token gates
✅ User without a key
✅ Member with expired key
✅ Member with a valid key
✅ Chains tested: Polygon & Goerli

### CODE REVIEW

File: connectors/chains.ts [hunk 1]
```diff
@@ -1,5 +1,6 @@
 import { log } from '@charmverse/core/log';
 // ref: https://wagmi.sh/core/chains
+import networks from '@unlock-protocol/networks';
```
Comment (mattcasey) [inline connectors/chains.ts:3]: it doesn't look like we need this to define our chain config at all, it just ends up exposting 'unlockChains'. It's a grey area but just to reduce file size everywhere, it might be a good idea to put this config inside of lib/tokenGates/unlock instead

File: lib/tokenGates/unlock/getLockMetadata.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import { LocksmithService } from '@unlock-protocol/unlock-js';
```
Comment (mattcasey) [inline lib/tokenGates/unlock/getLockMetadata.ts:1]: this is a really massive lib, actually! VS Code says 1.4mb. it would be nice if we didn't need it, since our dev env is already pretty bloated

File: lib/tokenGates/unlock/getLockMetadata.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+import type { GetLockPayload } from './getLockDetails';
+
+export async function getLockMetadata(values: Pick<GetLockPayload, 'chainId' | 'contract'>) {
```
Comment (mattcasey) [inline lib/tokenGates/unlock/getLockMetadata.ts:5]: i wouldn't have reused the input type here from getLockDetails. For one, both files are importing from each other which is hard to follow. Second, there's no reason the interface to getLockMetadata needs to be the same as getLockDetails. In fact, it requires you to cast chainId to Number twice.

File: lib/tokenGates/validateTokenGateConditions.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+const UnlockProtocolSchema = yup.object().shape({
+  contract: yup
+    .string()
+    .required()
+    .test('isAddress', 'Invalid address', (value) => isValidChainAddress(value)),
+  chainId: yup.number().required()
+});
+
+const LitProtocolSchema = yup.object().shape({
```
Comment (mattcasey) [inline lib/tokenGates/validateTokenGateConditions.ts:15]: nice, i like that you used yup instead of writing the rules yourself

Comment (mattcasey) [reply lib/tokenGates/validateTokenGateConditions.ts:15]: would be interesting to see some tests on this

File: lib/tokenGates/verifyTokenGates.ts [hunk 5]
```diff
@@ -49,28 +49,40 @@ export async function verifyTokenGates({ spaceId, userId, tokens }: Props): Prom
...
+                verified: true,
+                grantedRoles: matchingTokenGate.tokenGateToRoles.map((tgr) => tgr.role.id)
+              };
+            }
           }
+        } else {
+          return {
+            ...matchingTokenGate,
+            jwt: tk.signedToken,
+            verified: true,
```
Comment (mattcasey) [inline lib/tokenGates/verifyTokenGates.ts:83]: is this a bug? don't we need to still verify the gate thru unlock?

File: pages/api/token-gates/evaluate.ts [hunk 6]
```diff
@@ -1,21 +1,22 @@
...
-async function verifyWallet(req: NextApiRequest, res: NextApiResponse) {
-  const user = req.session.user;
+async function verifyWallet(req: NextApiRequest, res: NextApiResponse<TokenGateEvaluationResult>) {
+  const body = req.body as TokenGateEvaluationAttempt;
 
-  const result = await evaluateTokenGateEligibility({
-    ...req.body,
-    userId: user.id
-  });
+  const result = await evaluateTokenGateEligibility(body);
```
Comment (mattcasey) [inline pages/api/token-gates/evaluate.ts:19]: interesting, we were not even using userId it seems

File: charmClient/apis/tokenGates.ts [hunk 7]
```diff
@@ -1,44 +1,8 @@
...
-  }
-
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
-  verifyTokenGate(
-    verification: Omit<TokenGateVerificationRequest, 'userId'>
-  ): Promise<{ error?: string; success?: boolean }> {
-    return http.POST('/api/token-gates/verify', verification);
-  }
-
-  evaluateTokenGateEligibility(
```
Comment (mattcasey) [inline charmClient/apis/tokenGates.ts:36]: thanks for cleaning up

File: pages/api/token-gates/index.ts [hunk 8]
```diff
@@ -98,23 +84,22 @@ async function getTokenGates(req: NextApiRequest, res: NextApiResponse<TokenGate
...
     }
-  });
-  // Add identifiable names to token gates
-  const updatedResult = await updateTokenGateDetails(result as TokenGateWithRoles[]);
+  })) as TokenGateWithRoles[];
+
+  const tokenGatesWithDetails = await updateTokenGatesDetails(result);
 
-  res.status(200).json(updatedResult);
+  res.status(200).json(tokenGatesWithDetails);
```
Comment (mattcasey) [inline pages/api/token-gates/index.ts:102]: It looks like we don't even use the response in the UI. I think this could just be res.status(200).end()

File: charmClient/hooks/helpers.ts [hunk 9]
```diff
@@ -34,6 +34,14 @@ export function usePUT<T, U = unknown>(path: string) {
   });
 }
 
+// To be used when you need to trigger a get request on demand
+export function useGETTrigger<T, U = unknown>(path: string) {
```
Comment (mattcasey) [inline charmClient/hooks/helpers.ts:38]: i needed this! thanks. One suggestion is to just call it useGETtrigger. I dont think trigger needs to be capitalized

File: lib/tokenGates/evaluateEligibility.ts [hunk 10]
```diff
@@ -153,3 +120,70 @@ export async function evaluateTokenGateEligibility({
...
+
+export async function getTokenGateResults(tokenGate: TokenGateWithRoles, authSig: AuthSig) {
+  if (tokenGate.type === 'unlock') {
+    return getUnlockProtocolTokenGateResults(tokenGate, authSig);
+  } else {
+    return getLitTokenGateResults(tokenGate, authSig);
+  }
+}
+
+export async function getUnlockProtocolTokenGateResults(tokenGate: TokenGateWithRoles<'unlock'>, authSig: AuthSig) {
```
Comment (mattcasey) [inline lib/tokenGates/evaluateEligibility.ts:132]: dont think this needs to be exported

File: lib/tokenGates/evaluateEligibility.ts [hunk 11]
```diff
@@ -123,7 +90,7 @@ export async function evaluateTokenGateEligibility({
     })
   );
 
-  const successGates = tokenGateResults.filter((result) => result !== null) as TokenGateJwt[];
+  const successGates = tokenGateResults.filter(isTruthy);
```
Comment (mattcasey) [inline lib/tokenGates/evaluateEligibility.ts:93]: nice

File: pages/api/token-gates/evaluate.ts (hunk 6 above)
Comment (valentinludu) [reply pages/api/token-gates/evaluate.ts:19]: yes, because we had req.body as any, typescript didn't throw an error

File: charmClient/apis/tokenGates.ts (hunk 7 above)
Comment (valentinludu) [reply charmClient/apis/tokenGates.ts:36]: Sure. It's nice now that we don't have to create 3 states for data, loading and error in the frontend and we get everything out of the box from swr

File: charmClient/hooks/helpers.ts (hunk 9 above)
Comment (valentinludu) [reply charmClient/hooks/helpers.ts:38]: yeah, it looks weird with TT

File: lib/tokenGates/verifyTokenGates.ts (hunk 5 above)
Comment (valentinludu) [reply lib/tokenGates/verifyTokenGates.ts:83]: I will think of a better solution here

File: lib/tokenGates/validateTokenGateConditions.ts (hunk 4 above)
Comment (valentinludu) [reply lib/tokenGates/validateTokenGateConditions.ts:15]: I know we discussed about validating our endpoints payload so this can be useful in the future.

File: pages/api/token-gates/index.ts (hunk 8 above)
Comment (valentinludu) [reply pages/api/token-gates/index.ts:102]: This is the GET method. I think you are referring to the POST method. Yes, I will remove the response.

File: lib/tokenGates/evaluateEligibility.ts (hunk 10 above)
Comment (valentinludu) [reply lib/tokenGates/evaluateEligibility.ts:132]: Done. I removed export from 3 func that were not exported.

File: lib/tokenGates/unlock/getLockMetadata.ts (hunk 3 above)
Comment (valentinludu) [reply lib/tokenGates/unlock/getLockMetadata.ts]: Yep. I remove id. I am checking outside if it is a number or  not and then I am sending it to the getLockMetadata

File: lib/tokenGates/verifyTokenGates.ts (hunk 5 above)
Comment (valentinludu) [reply lib/tokenGates/verifyTokenGates.ts]: I added the verification of unlock token gates here

**Review by mattcasey — APPROVED** [review]
